### PR TITLE
Refactor memcpy_async for easier extensions.

### LINKED ISSUE #57: [FEA]: Redesign `memcpy_async` implementation to support more general src/dest dispatch

### PR DESCRIPTION
**Pull request by griwes**
Also make the memcpy_async tests slightly more robust.

## Description

Closes #57 

## Checklist

- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: libcudacxx/include/cuda/std/detail/libcxx/include/__config [hunk 1]
```diff
@@ -55,6 +55,9 @@
 #define _LIBCUDACXX_COMPILER_NVCC_BELOW_11_3
 #endif // defined(_LIBCUDACXX_COMPILER_NVCC) && _LIBCUDACXX_CUDACC_VER < 1103000
 
+#if defined(_LIBCUDACXX_COMPILER_NVCC) && _LIBCUDACXX_CUDACC_VER < 1107000
+#define _LIBCUDACXX_COMPILER_NVCC_BELOW_11_7
+#endif // defined(_LIBCUDACXX_COMPILER_NVCC) && _LIBCUDACXX_CUDACC_VER < 1107000
```
Comment (miscco) [inline libcudacxx/include/cuda/std/detail/libcxx/include/__config:60]: There was the fix from @wmaxey that we should change that to
```suggestion
#if defined(__CUDACC__) && _LIBCUDACXX_CUDACC_VER < 1107000
#define _LIBCUDACXX_COMPILER_CUDACC_BELOW_11_7
#endif // defined(_LIBCUDACXX_COMPILER_NVCC) && _LIBCUDACXX_CUDACC_VER < 1107000
```

File: libcudacxx/include/cuda/pipeline [hunk 2]
```diff
@@ -514,59 +516,45 @@ _LIBCUDACXX_BEGIN_NAMESPACE_CUDA
...
-        // incorrectly classified as not trivially copyable. Remove this assertion to allow for their usage with
-        // memcpy_async when compiling with GCC 4.8.
-        // FIXME: remove the #if once GCC 4.8 is no longer supported.
-    #if !defined(_LIBCUDACXX_COMPILER_GCC) || _GNUC_VER > 408
-        static_assert(std::is_trivially_copyable<_Type>::value, "memcpy_async requires a trivially copyable type");
-    #endif
-
-        return __memcpy_async<alignof(_Type)>(__group, reinterpret_cast<char *>(__destination), reinterpret_cast<char const *>(__source), __size, __pipeline);
-    }
+    template<thread_scope _Sco, bool _Tx, typename _Arch, __space _OutSpace, __space _InSpace, __space _SyncSpace>
```
Comment (miscco) [inline libcudacxx/include/cuda/pipeline:523]: I really want to get away from boolean template parameters when possible. 

It is hard to follow what "false" means. I would greatly prefer if we had a `enum class _UseMeow : bool { __no, __yes}` and pass that around

File: libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/barrier.h [hunk 3]
```diff
@@ -590,211 +574,92 @@ _LIBCUDACXX_INLINE_VISIBILITY constexpr bool __unused(_Ty...) {return true;}
...
+    _LIBCUDACXX_INLINE_VISIBILITY
+    void operator()(_Arch) {
+        const void * __p = &__b;
+
+        _LIBCUDACXX_HANDLE_POINTER_SPACE(__p,
+            __memcpy_async_sync_hooks<
+                barrier<_Sco, _CompF>, false, _Arch,
+                __space::__shared, __space::__global, __p_space_t::value
+            >::__synchronize(_Arch{}, __b,
+                __arch::__is_cuda_provides_sm<_Arch, 80>::value ? async_contract_fulfillment::async : async_contract_fulfillment::none);
```
Comment (miscco) [inline libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/barrier.h:591]: As said above I believe this should be
```suggestion
        _LIBCUDACXX_HANDLE_POINTER_SPACE(__p, 
            __memcpy_async_sync_hooks<_Tx::__no, __space::__shared, __space::__global, __p_space_t::value
            >::__synchronize(_Arch{}, __b,
                __arch::__is_cuda_provides_sm<_Arch, 80>::value ? async_contract_fulfillment::async : async_contract_fulfillment::none);
```

Comment (miscco) [inline libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/barrier.h:591]: We should have a helper for that

File: libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h [hunk 4]
```diff
@@ -0,0 +1,525 @@
...
+};
+
+template<__space _Sp>
+using __space_constant = _CUDA_VSTD::integral_constant<__space, _Sp>;
+
+_LIBCUDACXX_DEVICE
+bool __is_cluster_shared(const void * __p) {
+    NV_IF_TARGET(NV_PROVIDES_SM_90,
+        (return __isClusterShared(__p);),
+        (return false;)
```
Comment (miscco) [inline libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h:63]: We might need to add a void cast for the argument to avoid compiler warnings
```suggestion
        ((void)__p; return false;)
```

File: libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h [hunk 5]
```diff
@@ -0,0 +1,525 @@
...
+
+#define _LIBCUDACXX_HANDLE_POINTER_SPACE(name, ...) \
+    NV_IF_ELSE_TARGET( \
+        NV_IS_DEVICE, ( \
+            _LIBCUDACXX_HANDLE_SPACE(__is_cluster_shared,__cluster, name, __VA_ARGS__) \
+            _LIBCUDACXX_HANDLE_SPACE(__isShared, __shared, name, __VA_ARGS__) \
+            _LIBCUDACXX_HANDLE_SPACE(__isGlobal, __global, name, __VA_ARGS__) \
+            _LIBCUDACXX_HANDLE_SPACE(__is_grid_constant, __grid_constant, name, __VA_ARGS__) \
+            _LIBCUDACXX_HANDLE_SPACE(__isConstant, __constant, name, __VA_ARGS__) \
+            _LIBCUDACXX_HANDLE_SPACE(/* __isLocal when not broken */ bool, __local, name, __VA_ARGS__) \
```
Comment (miscco) [inline libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h:90]: loving that comment

File: libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h [hunk 6]
```diff
@@ -0,0 +1,525 @@
...
+
+struct __alignment_tag {};
+template<_CUDA_VSTD::size_t _Alignment>
+using __alignment = __down_convertible_constant<__alignment_tag, _Alignment>;
+
+template<_CUDA_VSTD::size_t _GuaranteedAlignment,
+    _CUDA_VSTD::size_t _MaxInterestingAlignment,
+    _CUDA_VSTD::size_t _MinInterestingAlignment,
+    _CUDA_VSTD::size_t _Alignment,
+    typename = void>
```
Comment (miscco) [inline libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h:117]: Please align

File: libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h [hunk 7]
```diff
@@ -0,0 +1,525 @@
...
+struct __invoke_if_applicable {
+    template<typename _Fn>
+    _LIBCUDACXX_INLINE_VISIBILITY static async_contract_fulfillment __invoke(_Fn && __f) {
```
Comment (miscco) [inline libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h:120]: As a reader the static and the return type are the most important parts here, so please add a line break after the macro
```suggestion
    _LIBCUDACXX_INLINE_VISIBILITY
    static async_contract_fulfillment __invoke(_Fn && __f) {
```

File: libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h [hunk 8]
```diff
@@ -0,0 +1,525 @@
...
+    _CUDA_VSTD::size_t _MinInterestingAlignment,
+    _CUDA_VSTD::size_t _Alignment>
+struct __invoke_if_applicable<_GuaranteedAlignment,
+    _MaxInterestingAlignment,
+    _MinInterestingAlignment,
+    _Alignment,
+    _CUDA_VSTD::__enable_if_t<
+        _Alignment >= _MaxInterestingAlignment
+            || _Alignment < _MinInterestingAlignment
+            || _Alignment < _GuaranteedAlignment
```
Comment (miscco) [inline libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h:136]: Can we move those conditions into some `_LIBCUDACXX_INLINE_VAR constexpr bool descriptiveName`

File: libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h [hunk 9]
```diff
@@ -0,0 +1,525 @@
...
+        _LIBCUDACXX_UNREACHABLE();
+    }
+};
+
+template<_CUDA_VSTD::size_t _MaxInterestingAlignment,
+    _CUDA_VSTD::size_t _MinInterestingAlignment,
+    _CUDA_VSTD::size_t _GuaranteedAlignment,
+    typename _Fn>
+_LIBCUDACXX_INLINE_VISIBILITY async_contract_fulfillment __dispatch_alignment_bit(_Fn && __f, _CUDA_VSTD::size_t __alignment_fsb) {
+    _CUDA_VSTD::size_t __alignment_v = 1ull << (__alignment_fsb - 1);
```
Comment (miscco) [inline libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h:149]:
```suggestion
    const _CUDA_VSTD::size_t __alignment_v = 1ull << (__alignment_fsb - 1);
```

File: libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h [hunk 10]
```diff
@@ -0,0 +1,525 @@
...
+
+    if (__alignment_v >= _MaxInterestingAlignment) [[likely]] {
```
Comment (miscco) [inline libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h:151]: I am not sure about those attributes. I fear we will need to _UGLIFY them via a macro

File: libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h [hunk 11]
```diff
@@ -0,0 +1,525 @@
...
+        _ADD_CASE(2);
+        _ADD_CASE(1);
+
+#undef _ADD_CASE
+    }
+
+    _LIBCUDACXX_UNREACHABLE();
+}
+
+_LIBCUDACXX_INLINE_VISIBILITY _CUDA_VSTD::size_t __get_size(_CUDA_VSTD::size_t __size) {
```
Comment (miscco) [inline libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h:187]: Why are those functions not `constexpr`?

File: libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h [hunk 12]
```diff
@@ -0,0 +1,525 @@
...
+};
+}
+
+_LIBCUDACXX_INLINE_VISIBILITY
+async_contract_fulfillment __strided_memcpy(_CUDA_VSTD::size_t __rank, _CUDA_VSTD::size_t __group_size, char * __out_ptr, const char * __in_ptr, _CUDA_VSTD::size_t __size, _CUDA_VSTD::size_t __alignment) {
+    if (__group_size == 1) {
+        memcpy(__out_ptr, __in_ptr, __size);
+    }
+    else {
+        for (_CUDA_VSTD::size_t __offset = __rank * __alignment; __offset < __size; __offset += __group_size * __alignment) {
```
Comment (miscco) [inline libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h:227]: Why is `__offset` not starting at 0 ?

File: libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h [hunk 13]
```diff
@@ -0,0 +1,525 @@
...
+    _Size __size;
+    _SyncObject & __sync;
+
+    template<typename _Arch>
+    _LIBCUDACXX_INLINE_VISIBILITY
+    async_contract_fulfillment operator()(_Arch) {
+        auto __f = __memcpy_async_space_dispatcher<_Tx, _NativeAlignment, _Arch>(_CUDA_VSTD::forward<_Group>(__g), __out_ptr, __in_ptr, __size, __sync);
+
+        auto __sync_ptr = &__sync;
+        _LIBCUDACXX_HANDLE_POINTER_SPACE(__out_ptr, _LIBCUDACXX_HANDLE_POINTER_SPACE(__in_ptr, _LIBCUDACXX_HANDLE_POINTER_SPACE(__sync_ptr,
```
Comment (miscco) [inline libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h:491]: 🙀

File: libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h [hunk 14]
```diff
@@ -0,0 +1,525 @@
...
+    _Size __size,
+    _SyncObject & __sync)
+{
+    auto __sync_ptr = &__sync;
+    return __dispatch_architecture<async_contract_fulfillment>(__memcpy_async_arch_dispatcher<_Tx, _NativeAlignment>(_CUDA_VSTD::forward<_Group>(__g), __out_ptr, __in_ptr, __size, __sync));
+}
+
+_LIBCUDACXX_END_NAMESPACE_CUDA
+
+#endif
```
Comment (miscco) [inline libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h:525]: Missing comment

**Review by miscco — COMMENTED** [review]
First pass review

File: libcudacxx/include/cuda/pipeline (hunk 2 above)
Comment (griwes) [reply libcudacxx/include/cuda/pipeline:523]: I'm okay with that.

File: libcudacxx/include/cuda/pipeline [hunk 15]
```diff
@@ -514,59 +516,45 @@ _LIBCUDACXX_BEGIN_NAMESPACE_CUDA
...
+    struct __memcpy_async_sync_hooks<
+        pipeline<_Sco>, _Tx, _Arch,
+        _OutSpace, _InSpace, _SyncSpace
```
[the remark this answers is not in the document]
Comment (griwes) [reply libcudacxx/include/cuda/pipeline:526]: For sync_hooks specifically, and for _Arch specifically, maybe? I can experiment with that. _Sco is an inherent part of the pipeline type, so no on that one. Also no on the same question applied to the main hooks, because there the arch being matched early is fundamental to how I'm eliminating *a lot* of template instantiations and generating code (before I did that, I'd have ptxas take upwards of 50 gigs of memory trying to make sense out of the thing I was spewing out).

File: libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/barrier.h (hunk 3 above)
Comment (griwes) [reply libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/barrier.h:591]: It definitely needs to know at least the exact type (because I want per-type specializations to be local to that type). I'll experiment with the arch part.

Comment (griwes) [reply libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/barrier.h:591]: This is literally the only place where this pattern is used, and it's only because of an artifact of how I'm reusing almost-unrelated code to satisfy pipeline's needs.

Comment (griwes) [reply libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/barrier.h:591]: You mean the expansion of the macro? I've tried that, it ends up being messier and harder on the compiler. I've started with a helper, and then with this form within a helper, and eventually ended up going to just straight macro invocations, because they are much easier to work with (especially in the case with many arguments, as in the main feature of this PR).

File: libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h (hunk 6 above)
Comment (griwes) [reply libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h:117]: Please elaborate, I'm unsure what you mean.

File: libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h (hunk 8 above)
Comment (griwes) [reply libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h:136]: We could, but I think I'd rather just write a comment here if we think that's more useful? The descriptive name would just be a weird restating of the name of this class template, so I don't know if it'd actually help.

File: libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h (hunk 10 above)
Comment (griwes) [reply libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h:151]: Yeah, I have rewritten it with `__builtin_expect` for now, I do believe we'll need a proper macro for this.

File: libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h (hunk 11 above)
Comment (griwes) [reply libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h:187]: I have a local patch that turns them into not-even-actual-functions, so this isn't relevant anymore.

File: libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h (hunk 12 above)
Comment (griwes) [reply libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h:227]: This is handling cooperative copy; the thread with rank 0 will start at 0, the thread with rank 1 will start at 1, and so on. This way we have good locality of the data that threads in a specific warp are moving, which means the hardware can coalesce it. This same idea is also done with `cp.async` later down the line.

File: libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h (hunk 13 above)
Comment (griwes) [reply libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h:491]: Trust me, this is easier to work with than a wrapper that I started with.

File: libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h (hunk 6 above)
Comment (miscco) [reply libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h:117]: In the editor it looked like the alignment could be 2 spaces to the right to properly align all arguments

File: libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/barrier.h (hunk 3 above)
Comment (miscco) [reply libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/barrier.h:591]: Oh no I meant:
```cpp
__arch::__is_cuda_provides_sm<_Arch, 80>::value ? async_contract_fulfillment::async : async_contract_fulfillment::none
```

File: libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h (hunk 8 above)
Comment (miscco) [reply libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h:136]: Yeah fine by me, if there are more of them we can still refactor it out

File: libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/barrier.h (hunk 3 above)
Comment (griwes) [reply libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/barrier.h:591]: If that ever shows up elsewhere, then we've done something wrong - this is very much just a deep detail of the interactions between pipeline and barrier.

File: libcudacxx/include/cuda/pipeline (hunk 15 above)
[the remark this answers is not in the document]
Comment (griwes) [reply libcudacxx/include/cuda/pipeline:526]: I think this would make it annoying to do the "fallback" for barrier, because the fallback happens for both non-sm80+ arches, but also for non-shared spaces; as currently spelled out, sure, there's a bit of sfinae, but then we explicitly inherit from `__noop_sync_traits`, which I think is a self documenting piece of code. Without the arch in the struct's template parameter list, we'd now kinda have two places that do that fallback, and the global->shared spec would also need to inherit from `__noop_sync_traits`, which I think is awkward.

[the remark this answers is not in the document]
Comment (miscco) [reply libcudacxx/include/cuda/pipeline:526]: Sounds good

File: libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h [hunk 16]
```diff
@@ -0,0 +1,575 @@
...
+{
+    NV_DISPATCH_TARGET(
+        NV_PROVIDES_SM_90,
+        (return _CUDA_VSTD::forward<_Fn>(__f)(__arch::__cuda<90>());),
+        NV_PROVIDES_SM_80,
+        (return _CUDA_VSTD::forward<_Fn>(__f)(__arch::__cuda<80>());),
+        NV_PROVIDES_SM_70,
+        (return _CUDA_VSTD::forward<_Fn>(__f)(__arch::__cuda<70>());),
+        NV_IS_HOST,
+        (return _CUDA_VSTD::forward<_Fn>(__f)(__arch::__host());))
```
Comment (miscco) [inline libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h:460]: Could you move the closing brace to a separate line, makes it easier to parse

File: libcudacxx/.upstream-tests/test/support/cuda_space_selector.h [hunk 17]
```diff
@@ -102,23 +103,23 @@ struct init_initializer {
...
+        cuda::std::__construct_at(&t, cuda::std::forward<Ts>(ts)...);
     }
 };
 
 struct default_initializer {
     template<typename T>
     __host__ __device__
     static void construct(T & t) {
-        new ((void*)&t) T;
+        ::new ((void*)&t) T;
```
[the remark this answers is not in the document]
Comment (griwes) [reply libcudacxx/.upstream-tests/test/support/cuda_space_selector.h:122]: Does that end up doing default-init or value-init?

File: libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/barrier.h [hunk 18]
```diff
@@ -145,7 +126,12 @@ friend class _CUDA_VSTD::__barrier_poll_tester_parity;
     }
 
     _LIBCUDACXX_INLINE_VISIBILITY
-    friend void init(barrier * __b, _CUDA_VSTD::ptrdiff_t __expected, _CUDA_VSTD::__empty_completion __completion = _CUDA_VSTD::__empty_completion()) {
+    friend void init(barrier * __b, _CUDA_VSTD::ptrdiff_t __expected) {
+        init(__b, __expected, _CUDA_VSTD::__empty_completion());
+    }
+
+    _LIBCUDACXX_INLINE_VISIBILITY
+    friend void init(barrier * __b, _CUDA_VSTD::ptrdiff_t __expected, _CUDA_VSTD::__empty_completion __completion) {
```
[the remark this answers is not in the document]
Comment (griwes) [reply libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/barrier.h:134]: No, because of complicated overload resolution rules. I needed to split this, because it became a template. You'll see that the init for the primary template of `cuda::barrier` also does this, and I don't remember exactly which rule governs this, just that I immediately remembered this happened before when I got stupid overload resolution errors this time around too.

File: libcudacxx/.upstream-tests/test/support/overrun_guard.h [hunk 19]
```diff
@@ -0,0 +1,75 @@
...
+// SPDX-FileCopyrightText: Copyright (c) 2023 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef TEST_OVERRUN_GUARD
+#define TEST_OVERRUN_GUARD
+
+#include <cuda/std/type_traits>
+
+template<typename T>
```
[the remark this answers is not in the document]
Comment (griwes) [reply libcudacxx/.upstream-tests/test/support/overrun_guard.h:16]: At some point during this work I managed to do a dumb and write over some bytes beyond the variable, which results in a silly behavior of an endless hang if you happen to overwrite the barrier; fun times. Hopefully this will catch the more likely cases of off-by-ones (that's what I did, didn't subtract 1 from the first set bit index when turning that index into an actual alignment value) with a reasonable assert message instead of a hang in `arrive_and_wait`.

File: libcudacxx/.upstream-tests/test/support/cuda_space_selector.h (hunk 17 above)
[the remark this answers is not in the document]
Comment (miscco) [reply libcudacxx/.upstream-tests/test/support/cuda_space_selector.h:122]: value initialization, as it calls T()

File: libcudacxx/.upstream-tests/test/support/overrun_guard.h (hunk 19 above)
[the remark this answers is not in the document]
Comment (miscco) [reply libcudacxx/.upstream-tests/test/support/overrun_guard.h:16]: Oh I mean in the file for future reference ;)

File: libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/barrier.h (hunk 18 above)
[the remark this answers is not in the document]
Comment (miscco) [reply libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/barrier.h:134]: :sad trombone noises:

File: libcudacxx/.upstream-tests/test/support/cuda_space_selector.h (hunk 17 above)
[the remark this answers is not in the document]
Comment (griwes) [reply libcudacxx/.upstream-tests/test/support/cuda_space_selector.h:122]: That'd change the semantics, this explicitly wants to do default init.

File: libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/barrier.h [hunk 20]
```diff
@@ -557,244 +543,103 @@ _LIBCUDACXX_END_NAMESPACE_CUDA
 _LIBCUDACXX_BEGIN_NAMESPACE_CUDA_DEVICE
 
 _LIBCUDACXX_DEVICE
-inline _CUDA_VSTD::uint64_t * barrier_native_handle(barrier<thread_scope_block> & b) {
-    return reinterpret_cast<_CUDA_VSTD::uint64_t *>(&b.__barrier);
+inline _CUDA_VSTD::uint64_t * barrier_native_handle(barrier<thread_scope_block> & __b) {
```
[the remark this answers is not in the document]
Comment (griwes) [reply libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/barrier.h:546]: We can, though that's going to create more issue/PR noise :P

File: libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h [hunk 21]
```diff
@@ -0,0 +1,575 @@
...
+};
+
+template<_CUDA_VSTD::size_t _MaxInterestingAlignment,
+    _CUDA_VSTD::size_t _MinInterestingAlignment,
+    _CUDA_VSTD::size_t _GuaranteedAlignment,
+    typename _Fn>
+_LIBCUDACXX_INLINE_VISIBILITY async_contract_fulfillment __dispatch_alignment_bit(_Fn && __f, _CUDA_VSTD::size_t __alignment_fsb) {
+    const _CUDA_VSTD::size_t __alignment_v = 1ull << (__alignment_fsb - 1);
+
+    if (__builtin_expect(__alignment_v >= _MaxInterestingAlignment, true)) {
```
[the remark this answers is not in the document]
Comment (griwes) [reply libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h:184]: All things currently in the CI matrix work fine with this. We need to run this on Windows though, it's ~~possible~~likely that MSVC doesn't like it.

File: libcudacxx/include/cuda/pipeline [hunk 22]
```diff
@@ -514,59 +516,45 @@ _LIBCUDACXX_BEGIN_NAMESPACE_CUDA
...
-        // incorrectly classified as not trivially copyable. Remove this assertion to allow for their usage with
-        // memcpy_async when compiling with GCC 4.8.
-        // FIXME: remove the #if once GCC 4.8 is no longer supported.
-    #if !defined(_LIBCUDACXX_COMPILER_GCC) || _GNUC_VER > 408
-        static_assert(std::is_trivially_copyable<_Type>::value, "memcpy_async requires a trivially copyable type");
-    #endif
-
-        return __memcpy_async<alignof(_Type)>(__group, reinterpret_cast<char *>(__destination), reinterpret_cast<char const *>(__source), __size, __pipeline);
-    }
+    template<thread_scope _Sco, __tx_api _Tx, typename _Arch, __space _OutSpace, __space _InSpace, __space _SyncSpace>
```
[the remark this answers is not in the document]
Comment (griwes) [reply libcudacxx/include/cuda/pipeline:523]: I really dislike `class`, because it's the less semantically accurate of the two spellings (an argument could be an `int`, and `int` is definitely not a class) - but we have both throughout the library, so we should probably settle on a policy and do a library-wide unification if we want to.

File: libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h [hunk 23]
```diff
@@ -316,27 +317,39 @@ struct __memcpy_async_default_aligned_impl<
     }
 };
 
+_LIBCUDACXX_INLINE_VISIBILITY
+_CUDA_VSTD::size_t __memcpy_async_ffs(_CUDA_VSTD::size_t __val) {
+    NV_IF_ELSE_TARGET(
+        NV_IS_DEVICE,
+        (return __ffsll(__val);),
+        (return _CUDA_VSTD::__libcpp_ctz(__val) + 1;)
```
Comment (wmaxey) [inline libcudacxx/include/cuda/std/detail/libcxx/include/__cuda/memcpy_async.h:379]: I doubt it might be more efficient, but `countr_zero` will use the right intrinsic depending on the context. Possibly even using the host's builtin for constexpr evaluation.

**Review by miscco — COMMENTED** [review]
I was looking through PRs and realized I did not send the review 🤦‍♂️

**Comment by jrhemstad** [comment]
I think we can just go ahead and close this PR for now. It's not likely to be revived any time soon. 

@griwes do you agree?
